File: Lv.1/ex33.js
//프로그래머스 나누어 떨어지는 숫자 배열
function solution(arr, divisor) {
    var answer = [];
    for(let i of arr){
        if(i%divisor ===0)
            answer.push(i);
    }
    
    if(answer.length === 0)
        return [-1];
    
    return answer.sort((a,b) => a-b);
}

// sort((a,b) => a-b); 없이 sort() 만하면 테스트케이스에 (10,5)는 오름차순으로 정렬되지않는다는걸 알게되었다.
// 검색결과 -> sort()는 기본적으로 문자열로 변환한 각 요소를 비교하여 정렬하므로 (10, 5)는 문자열로 변환된 후 각 자리의 첫 번째
// 숫자를 비교하고 그다음으로 넘어가기때문에 수를 정렬할때는 정렬함수를 지정해서 오름차순, 내림차순을 정해줘야한다.
function solution(arr, divisor) {
    var answer = [];
    arr.map((o) => {
        o % divisor === 0 && answer.push(o);
    })
    return answer.length ? answer.sort((a, b) => a - b) : [-1];

}
//이거는 별 특별한 코드는 아니지만, o % divisor === 0 && answer.push(o); 부분이 특이해서 가져왔다.
//map에서 o % divisor === 0 이 참이라면 &&의 뒤에부분은 무조건 실행된다고한다

File: Lv.1/ex33.java
//가장 가까운 같은 글자

//첫 글자는 무조건 -1일수밖에없으니 1부터 for문을 돌리고
//2중 for문으로 본인 바로 앞 글자까지 매번 검사하며 자신과 같은 글자가 나올때 해당 index를 저장한다.
//그후 index가 초기값인지를 구분한뒤 값을 처리하여 담아주면 끝.
public class ex33 {
    public int[] solution(String s) {
        int[] answer = new int[s.length()];
        answer[0] = -1;
        
        for(int i=1; i<s.length(); i++){
            int before_idx = -1;
            for(int j=0; j<i; j++){
                if(s.charAt(i) == s.charAt(j)){
                    before_idx = j;
                }
            }
            if(before_idx == -1){
                answer[i] = -1;
            }else{
                answer[i] = i-before_idx;
            }
        }
        
        return answer;
    }


    //다른 풀이 제일 마음에드는거.
    //substring으로 자기 자신 이저의 문자열을 잘라다가 indexOf 함수를 이용해 판별을 했다.
    public int[] solution2(String str) {
        int[] result = new int[str.length()];

        for(int i=0;i<str.length();i++){

            String subStr = str.substring(0,i);
            if(subStr.indexOf(str.charAt(i))==-1) {
                result[i] = -1;
            }else {
                result[i] = i-subStr.lastIndexOf(str.charAt(i));
            }
        }
        return result;
    }
}
